controller/configmap/synchandler.go

```diff
@@ -0,0 +1,105 @@
+package configmap
+
+/*
+Copyright 2020 Crunchy Data Solutions, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+
+	crv1 "github.com/crunchydata/postgres-operator/apis/crunchydata.com/v1"
+	"github.com/crunchydata/postgres-operator/config"
+	cfg "github.com/crunchydata/postgres-operator/operator/config"
+)
+
+// handleConfigMapSync is responsible for syncing a configMap resource that has obtained from
+// the ConfigMap controller's worker queue
+func (c *Controller) handleConfigMapSync(key string) error {
+
+	log.Debugf("ConfigMap Controller: handling a configmap sync for key %s", key)
+
+	namespace, configMapName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	configMap, err := c.cmLister.ConfigMaps(namespace).Get(configMapName)
+	if err != nil {
+		return err
+	}
+	clusterName := configMap.GetObjectMeta().GetLabels()[config.LABEL_PG_CLUSTER]
+
+	cluster, err := c.pgclusterLister.Pgclusters(namespace).Get(clusterName)
+	if err != nil && kerrors.IsNotFound(err) {
+		log.Debugf("ConfigMap Controller: cannot find pgcluster for configMap %s (namespace %s),"+
+			"ignoring", configMapName, namespace)
+	} else if err != nil {
+		return err
+	}
```

Comment:
So the real issue here is re-queuing, which is included in the new ConfigMap controller.  Basically, any errors that are returned from this function result in the item being re-queued onto the worker queue.  The idea here is to catch any transient errors, API server hiccups, etc. that will typically resolve themselves, allowing for successful processing of the item on a subsequent attempt (as it stands right now, its really only the calls to grab the various resources from the listers that can actually return errors, so it's pretty unlikely anything will actually be re-queued).

So in this sense I was trying to be very careful about what errors escape this function and result in a re-queue, and in the event a cluster was deleted and for some reason and the configMap remained, I didn't want to trigger re-queue after re-queue for a `pgcluster` that is gone and will therefore never exist (rate limiting does add protection, but I am trying to protect  it from getting even to that point here).  Also, I did think it was important to introduce this concept in general, i.e. that we need to consider transient errors in a cloud env and use the re-queuing mechanisms to attempt again and recover from those errors. 

Maybe logging an Error vs a Debug message is the answer here?  I'll move in that direction, but just let me know if any other thoughts.  I will also restructure per your suggestion with the embedded if.

Reply:
Hm..Would we then `return nil` if it's not found, so we don't process the rest of the function?

Reply:
Yea, agree that would definitely be cleaner.  Right now the logic immediately under it to check the cluster status would fail and return `nil` preventing an additional processing, but really it should just immediately return `nil` when this condition is encountered.

I'll go ahead an make the appropriate changes.

Reply:
Cool :+1:

Reply:
Code has been updated per recommendation.

Also, an error is now being logged to clearly indicate this is an error condition, even though re-queuing isn't desired (a PGHA `configMap` really should never exist without a `pgcluster`).